minibrowser: Add loading spinner

frereit opened the pull request:
I've started to implement a simple spinner that indicates when a page is loading. This PR is not complete yet but I'd like to get some feedback and discuss some implementation details if that's possible.

For this "MVP", I've simply added a new bool `loading_changed` analogous to the `history_changed` field in `PumpResult::Continue`. The `WebView` then keeps track of wether or not it is loading and exposes this via `WebView::loading()`, like `WebView::shutdown_requested`. The `Minibrowser` then gets a notification from the `App` when `loading` changed and updates the spinner accordingly.

A few questions arose while implementing this:

- The comments in the code and issue indicate that `EmbedderMsg::HeadParsed` should also "influence" the loading state, but I do not understand how this should surface. In my understanding, the spinner simply shows up when the document starts loading and stops once it is complete. What is the intended "intermediary" step here?

https://github.com/servo/servo/blob/8cfc6a1898d02eb1df9c13c3c410747cb1e0b412/ports/servoshell/webview.rs#L596-L598

- It seems a bit clunky to add a `xyz_changed` for everything about the `WebView` that could be required to bubble up to the `Minibrowser`. For example, showing the Favicon might also require a `favicon_changed: bool` in the future. Perhaps it would make sense to combine all these into a `toolbar_data_changed` or `metadata_changed` or something and let the `Minibrowser` figure out what it has to do (or just refresh the entire toolbar when anything changed, that shouldn't really impact performance). Is this something I should implement in this PR?

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #31689 

<!-- Either: -->
- [X] These changes do not require tests because it is purely visual in the minibrowser.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

frereit commented on ports/servoshell/minibrowser.rs:
```diff
@@ -314,4 +319,13 @@ impl Minibrowser {
+        &mut self,
+        browser: &mut WebViewManager<dyn WindowPortsMethods>,
+    ) -> bool {
+        self.loading = browser.loading();
+        return true;
```
Currently this always forces a `redraw`. I attempted to only redraw when `self.loading` changed like this:

```suggestion
        let need_redraw = self.loading != browser.loading();
        self.loading = browser.loading();
        return need_redraw;
```

But that resulted in some weird behavior where the spinner wasn't always being drawn. I'll investigate why this is the case but wanted to open the PR already to get feedback on the two questions in the other comments, I hope that's okay.

frereit replied:
This is resolved by the new commit, whch implements this without problems.

jdm commented:
>What is the intended "intermediary" step here?

One common pattern I've seen is a slow spinner in one direction while the initial network request is in flight, the the spinner changes direction or speeds up when the beginning of the response is received. This corresponds to the HeadParsed event.

delan commented:
> > What is the intended "intermediary" step here?
> 
> One common pattern I've seen is a slow spinner in one direction while the initial network request is in flight, then the spinner changes direction or speeds up when the beginning of the response is received. This corresponds to the HeadParsed event.

Yeah, HeadParsed happens somewhere between LoadStart and LoadComplete. It would be good to show these two periods differently, but if we can’t, then we should treat HeadParsed like LoadStart.

> It seems a bit clunky to add a `xyz_changed` for everything about the `WebView` that could be required to bubble up to the `Minibrowser`. For example, showing the Favicon might also require a `favicon_changed: bool` in the future. Perhaps it would make sense to combine all these into a `toolbar_data_changed` or `metadata_changed` or something and let the `Minibrowser` figure out what it has to do (or just refresh the entire toolbar when anything changed, that shouldn't really impact performance). Is this something I should implement in this PR?

Good point! Since we only use our history_changed flag to check if we should run an egui update (and maybe update some Minibrowser fields), I agree that it would be good to combine them.

frereit committed:
minibrowser: Rename "history_changed" flag to "need_update"

There are other data points in the toolbar that might need to be
updated. This commit prepares for that by generaliziing the
"history_changed" flag to a more generic "need_update" flag.

Signed-off-by: Frederik Reiter <hi@frereit.de>

frereit committed:
minibrowser: Add spinner to indicate loading status of the webview

Signed-off-by: Frederik Reiter <hi@frereit.de>

frereit commented:
> > > What is the intended "intermediary" step here?
> > 
> > 
> > One common pattern I've seen is a slow spinner in one direction while the initial network request is in flight, then the spinner changes direction or speeds up when the beginning of the response is received. This corresponds to the HeadParsed event.
> 
> Yeah, HeadParsed happens somewhere between LoadStart and LoadComplete. It would be good to show these two periods differently, but if we can’t, then we should treat HeadParsed like LoadStart.

Done. During the `LoadStart` -> `HeadParsed` period the spinner is now `Color32::GRAY` and the default white otherwise.

> > It seems a bit clunky to add a `xyz_changed` for everything about the `WebView` that could be required to bubble up to the `Minibrowser`. For example, showing the Favicon might also require a `favicon_changed: bool` in the future. Perhaps it would make sense to combine all these into a `toolbar_data_changed` or `metadata_changed` or something and let the `Minibrowser` figure out what it has to do (or just refresh the entire toolbar when anything changed, that shouldn't really impact performance). Is this something I should implement in this PR?
> 
> Good point! Since we only use our history_changed flag to check if we should run an egui update (and maybe update some Minibrowser fields), I agree that it would be good to combine them.

Done.

I've also changed the `webview` to always return `need_present: true` when the spinner should be active because the spinner is animated and requires regular redraws to not be choppy. Is this the idiomatic way to achieve this or should the redraw be requested some other way?

delan commented on ports/servoshell/minibrowser.rs:
```diff
@@ -314,4 +323,28 @@ impl Minibrowser {
+        &mut self,
+        browser: &mut WebViewManager<dyn WindowPortsMethods>,
+    ) -> bool {
+        let need_update = browser.load_status() == self.load_status;
```
should this be `!=`?

frereit replied:
You're right, thanks!

delan reviewed (commented):
Excellent work! One question…

delan commented:
> I've also changed the `webview` to always return `need_present: true` when the spinner should be active because the spinner is animated and requires regular redraws to not be choppy. Is this the idiomatic way to achieve this or should the redraw be requested some other way?

This makes sense and seems good enough to me.

delan reviewed (approved):
Great, thanks!

frereit commented:
@delan Did you mean to only enable auto-merge without triggering the CI run? Is there something I need to do to finish up this PR?

frereit commented:
Apologies for the format errors. Fixed.
